risc0/zkvm/src/guest/env/batcher.rs

```diff
@@ -15,10 +15,107 @@
 use alloc::vec;
 
 use anyhow::Result;
-use risc0_zkvm_platform::syscall::DIGEST_BYTES;
+use risc0_circuit_keccak::{KeccakState, KECCAK_CONTROL_ROOT, KECCAK_DEFAULT_PO2};
+use risc0_zkp::core::hash::sha::SHA256_INIT;
+use risc0_zkvm_platform::syscall::{
+    sys_keccak_permute, sys_prove_keccak, sys_sha_compress, DIGEST_BYTES, DIGEST_WORDS,
+};
 
 use crate::sha::Digest;
 
+const KECCAK_PERMUTE_CYCLES: usize = 200;
+const MAX_KECCAK_CYCLES: usize = 1 << KECCAK_DEFAULT_PO2;
+const MAX_KECCAK_INPUTS: usize = MAX_KECCAK_CYCLES / KECCAK_PERMUTE_CYCLES;
+
+/// This struct implements the batching of calls to the keccak accelerator.
+pub struct Keccak2Batcher {
+    claim_state: Digest,
+    inputs: Vec<KeccakState>,
+}
+
+impl Keccak2Batcher {
+    pub const fn init() -> Self {
+        Self {
+            claim_state: SHA256_INIT,
+            inputs: vec![],
+        }
+    }
+
+    pub fn update(&mut self, in_state: &KeccakState) -> KeccakState {
+        let mut out_state = KeccakState::default();
+        unsafe { sys_keccak_permute(in_state, &mut out_state) };
+        self.sha_single_keccak(in_state);
+        self.inputs.push(*in_state);
+        if self.inputs.len() == MAX_KECCAK_INPUTS {
+            // we've reached the limit. Create a proof request.
+            self.finalize();
+        }
+        out_state
+    }
+
+    pub fn finalize(&mut self) {
+        if self.inputs.is_empty() {
+            // no input so there's nothing to do
+            self.reset();
+            return;
+        }
+
+        let input: &[u32] = bytemuck::cast_slice(self.inputs.as_flattened());
+        let claim_digest = self.claim_digest();
+
+        unsafe {
+            sys_prove_keccak(
+                claim_digest.as_ref(),
+                KECCAK_DEFAULT_PO2,
+                KECCAK_CONTROL_ROOT.as_ref(),
+                input.as_ptr(),
+                input.len(),
+            );
+        }
+        crate::guest::env::verify_assumption(claim_digest, KECCAK_CONTROL_ROOT).unwrap();
+
+        self.reset();
+    }
+
+    fn claim_digest(&self) -> Digest {
+        let mut claim_digest = self.claim_state;
+        for word in claim_digest.as_mut_words().iter_mut() {
+            *word = word.to_be();
+        }
+        claim_digest
+    }
+
+    fn reset(&mut self) {
+        self.claim_state = SHA256_INIT;
+        self.inputs = vec![];
```

Comment:
```suggestion
        self.inputs.clear();
```

This will potentially avoid a re-alllocation.